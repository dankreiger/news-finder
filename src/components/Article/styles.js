import styled from 'styled-components';
import { Card, CardImg, CardImgOverlay } from 'reactstrap';

export const StyledCard = styled(Card)`
  position: relative;
  padding-bottom: ${(2 / 3) * 100}%;
  overflow: hidden;
  border: none;
  border-radius: 0;
  background-color: transparent;
  &:hover {
    img {
      transform: scale(1.01);
    }
    .card-img-overlay {
      padding-bottom: 40px;
    }
  }
`;

export const ArticleImage = styled(CardImg)`
  width: 100%;
  height: 100%;
  display: block;
  position: absolute;
  vertical-align: middle;
  object-fit: cover;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  transition: transform 250ms;
  border-radius: 0;
`;

export const CardOverlayStyled = styled(CardImgOverlay)`
  top: unset;
  height: 100%;
  display: flex;
  flex-direction: column;
  justify-content: flex-end;
  background: linear-gradient(to top, rgba(0, 0, 0, 0.9) 0, transparent);
  transition: padding-bottom 250ms;
`;
