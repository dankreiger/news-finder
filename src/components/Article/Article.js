import React from 'react';
import { string } from 'prop-types';

import ProgressiveImage from 'react-progressive-image';
import { CardTitle, CardText } from 'reactstrap';
import { ArticleImage, StyledCard, CardOverlayStyled } from './styles';

const Article = ({ url, urlToImage, title, description, sourceName }) => {
  return (
    <a
      className="d-block h-100"
      href={url}
      target="_blank"
      rel="noopener noreferrer"
    >
      <StyledCard className="h-100" inverse>
        <ProgressiveImage src={urlToImage} placeholder="./images/blur_img.jpg">
          {(src, loading) => (
            <ArticleImage
              style={{ opacity: loading ? 0.5 : 1 }}
              src={src}
              alt="an image"
            />
          )}
        </ProgressiveImage>

        <CardOverlayStyled>
          <CardTitle>{title}</CardTitle>
          <CardText className="articleDescription">{description}</CardText>
          <CardText>
            <small className="text-muted">{sourceName}</small>
          </CardText>
        </CardOverlayStyled>
      </StyledCard>
    </a>
  );
};

Article.propTypes = {
  url: string,
  urlToImage: string,
  title: string,
  description: string,
  sourceName: string
};

export default Article;
