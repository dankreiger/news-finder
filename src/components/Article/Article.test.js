import React from 'react';
import Article from './Article';
import { shallow } from 'enzyme';
import { mock4VisibleArticles } from '../../static/dummyData/mockData';
import _ from 'lodash';

describe('Article', () => {
  let component;
  let mockArticle = mock4VisibleArticles[0];
  const { url, urlToImage, title, description, source } = mockArticle;
  beforeEach(() => {
    component = shallow(
      <Article
        url={url}
        urlToImage={urlToImage}
        title={title}
        description={description}
        sourceName={source.name}
      />
    );
  });

  it('renders without crashing', () => {
    expect(component.length).toBe(1);
    expect(component).toMatchSnapshot();
  });

  it('renders a wrapping anchor tag with the article url prop as the href', () => {
    expect(component.find('a').prop('href')).toEqual(url);
  });

  it('renders a ProgressiveImage component with the article urlToImage prop as its image src', () => {
    expect(component.find('ProgressiveImage').prop('src')).toEqual(urlToImage);
  });

  it('renders a CardTitle h5 tag with the article title as its inner text', () => {
    expect(component.find('CardTitle').props().children).toEqual(title);
    expect(component.find('CardTitle').props().tag).toEqual('h5');
  });

  it('renders a CardText.articleDescription p tag with the article description as its inner text', () => {
    expect(component.find('.articleDescription').props().children).toEqual(
      description
    );
    expect(component.find('.articleDescription').props().tag).toEqual('p');
  });

  it('renders a text-muted span with the article source name', () => {
    expect(component.find('.text-muted').props().children).toEqual(source.name);
  });
});
