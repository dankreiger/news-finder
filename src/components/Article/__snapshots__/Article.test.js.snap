// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Article renders without crashing 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Article
    description="On Wednesday, Fox News shared one of the dumbest takes of our time about Jamie Lee Curtis, star of the Halloween re-boot. The conservative news outlet tweeted, \\"Jamie Lee Curtis wields firearms in new 'Halloween' movie despite advocating for gun control.\\" It'…"
    sourceName="Mashable"
    title="Fox News reveals it has no idea how movies work in a very misguided tweet about Jamie Lee Curtis"
    url="https://mashable.com/article/fox-news-jamie-lee-curtis-gun-control-tweet-halloween/"
    urlToImage="https://i.amz.mshcdn.com/CrqgZdv3yZ8ITPg9VA0qNEHVFfA=/1200x630/2018%2F10%2F10%2F6e%2F0d481eaee8564432b78be96cb96f1954.4fe76.jpg"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <Styled(Card)
        className="h-100"
        inverse={true}
      >
        <ProgressiveImage
          placeholder="./images/blur_img.jpg"
          src="https://i.amz.mshcdn.com/CrqgZdv3yZ8ITPg9VA0qNEHVFfA=/1200x630/2018%2F10%2F10%2F6e%2F0d481eaee8564432b78be96cb96f1954.4fe76.jpg"
        >
          [Function]
        </ProgressiveImage>
        <Styled(CardImgOverlay)>
          <CardTitle
            tag="h5"
          >
            Fox News reveals it has no idea how movies work in a very misguided tweet about Jamie Lee Curtis
          </CardTitle>
          <CardText
            className="articleDescription"
            tag="p"
          >
            On Wednesday, Fox News shared one of the dumbest takes of our time about Jamie Lee Curtis, star of the Halloween re-boot. The conservative news outlet tweeted, "Jamie Lee Curtis wields firearms in new 'Halloween' movie despite advocating for gun control." It'…
          </CardText>
          <CardText
            tag="p"
          >
            <small
              className="text-muted"
            >
              Mashable
            </small>
          </CardText>
        </Styled(CardImgOverlay)>
      </Styled(Card)>,
      "className": "d-block h-100",
      "href": "https://mashable.com/article/fox-news-jamie-lee-curtis-gun-control-tweet-halloween/",
      "rel": "noopener noreferrer",
      "target": "_blank",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <ProgressiveImage
            placeholder="./images/blur_img.jpg"
            src="https://i.amz.mshcdn.com/CrqgZdv3yZ8ITPg9VA0qNEHVFfA=/1200x630/2018%2F10%2F10%2F6e%2F0d481eaee8564432b78be96cb96f1954.4fe76.jpg"
          >
            [Function]
          </ProgressiveImage>,
          <Styled(CardImgOverlay)>
            <CardTitle
              tag="h5"
            >
              Fox News reveals it has no idea how movies work in a very misguided tweet about Jamie Lee Curtis
            </CardTitle>
            <CardText
              className="articleDescription"
              tag="p"
            >
              On Wednesday, Fox News shared one of the dumbest takes of our time about Jamie Lee Curtis, star of the Halloween re-boot. The conservative news outlet tweeted, "Jamie Lee Curtis wields firearms in new 'Halloween' movie despite advocating for gun control." It'…
            </CardText>
            <CardText
              tag="p"
            >
              <small
                className="text-muted"
              >
                Mashable
              </small>
            </CardText>
          </Styled(CardImgOverlay)>,
        ],
        "className": "h-100",
        "inverse": true,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": [Function],
            "placeholder": "./images/blur_img.jpg",
            "src": "https://i.amz.mshcdn.com/CrqgZdv3yZ8ITPg9VA0qNEHVFfA=/1200x630/2018%2F10%2F10%2F6e%2F0d481eaee8564432b78be96cb96f1954.4fe76.jpg",
          },
          "ref": null,
          "rendered": [Function],
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              <CardTitle
                tag="h5"
              >
                Fox News reveals it has no idea how movies work in a very misguided tweet about Jamie Lee Curtis
              </CardTitle>,
              <CardText
                className="articleDescription"
                tag="p"
              >
                On Wednesday, Fox News shared one of the dumbest takes of our time about Jamie Lee Curtis, star of the Halloween re-boot. The conservative news outlet tweeted, "Jamie Lee Curtis wields firearms in new 'Halloween' movie despite advocating for gun control." It'…
              </CardText>,
              <CardText
                tag="p"
              >
                <small
                  className="text-muted"
                >
                  Mashable
                </small>
              </CardText>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": "Fox News reveals it has no idea how movies work in a very misguided tweet about Jamie Lee Curtis",
                "tag": "h5",
              },
              "ref": null,
              "rendered": "Fox News reveals it has no idea how movies work in a very misguided tweet about Jamie Lee Curtis",
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": "On Wednesday, Fox News shared one of the dumbest takes of our time about Jamie Lee Curtis, star of the Halloween re-boot. The conservative news outlet tweeted, \\"Jamie Lee Curtis wields firearms in new 'Halloween' movie despite advocating for gun control.\\" It'…",
                "className": "articleDescription",
                "tag": "p",
              },
              "ref": null,
              "rendered": "On Wednesday, Fox News shared one of the dumbest takes of our time about Jamie Lee Curtis, star of the Halloween re-boot. The conservative news outlet tweeted, \\"Jamie Lee Curtis wields firearms in new 'Halloween' movie despite advocating for gun control.\\" It'…",
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": <small
                  className="text-muted"
                >
                  Mashable
                </small>,
                "tag": "p",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Mashable",
                  "className": "text-muted",
                },
                "ref": null,
                "rendered": "Mashable",
                "type": "small",
              },
              "type": [Function],
            },
          ],
          "type": [Function],
        },
      ],
      "type": [Function],
    },
    "type": "a",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <Styled(Card)
          className="h-100"
          inverse={true}
        >
          <ProgressiveImage
            placeholder="./images/blur_img.jpg"
            src="https://i.amz.mshcdn.com/CrqgZdv3yZ8ITPg9VA0qNEHVFfA=/1200x630/2018%2F10%2F10%2F6e%2F0d481eaee8564432b78be96cb96f1954.4fe76.jpg"
          >
            [Function]
          </ProgressiveImage>
          <Styled(CardImgOverlay)>
            <CardTitle
              tag="h5"
            >
              Fox News reveals it has no idea how movies work in a very misguided tweet about Jamie Lee Curtis
            </CardTitle>
            <CardText
              className="articleDescription"
              tag="p"
            >
              On Wednesday, Fox News shared one of the dumbest takes of our time about Jamie Lee Curtis, star of the Halloween re-boot. The conservative news outlet tweeted, "Jamie Lee Curtis wields firearms in new 'Halloween' movie despite advocating for gun control." It'…
            </CardText>
            <CardText
              tag="p"
            >
              <small
                className="text-muted"
              >
                Mashable
              </small>
            </CardText>
          </Styled(CardImgOverlay)>
        </Styled(Card)>,
        "className": "d-block h-100",
        "href": "https://mashable.com/article/fox-news-jamie-lee-curtis-gun-control-tweet-halloween/",
        "rel": "noopener noreferrer",
        "target": "_blank",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <ProgressiveImage
              placeholder="./images/blur_img.jpg"
              src="https://i.amz.mshcdn.com/CrqgZdv3yZ8ITPg9VA0qNEHVFfA=/1200x630/2018%2F10%2F10%2F6e%2F0d481eaee8564432b78be96cb96f1954.4fe76.jpg"
            >
              [Function]
            </ProgressiveImage>,
            <Styled(CardImgOverlay)>
              <CardTitle
                tag="h5"
              >
                Fox News reveals it has no idea how movies work in a very misguided tweet about Jamie Lee Curtis
              </CardTitle>
              <CardText
                className="articleDescription"
                tag="p"
              >
                On Wednesday, Fox News shared one of the dumbest takes of our time about Jamie Lee Curtis, star of the Halloween re-boot. The conservative news outlet tweeted, "Jamie Lee Curtis wields firearms in new 'Halloween' movie despite advocating for gun control." It'…
              </CardText>
              <CardText
                tag="p"
              >
                <small
                  className="text-muted"
                >
                  Mashable
                </small>
              </CardText>
            </Styled(CardImgOverlay)>,
          ],
          "className": "h-100",
          "inverse": true,
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": [Function],
              "placeholder": "./images/blur_img.jpg",
              "src": "https://i.amz.mshcdn.com/CrqgZdv3yZ8ITPg9VA0qNEHVFfA=/1200x630/2018%2F10%2F10%2F6e%2F0d481eaee8564432b78be96cb96f1954.4fe76.jpg",
            },
            "ref": null,
            "rendered": [Function],
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": Array [
                <CardTitle
                  tag="h5"
                >
                  Fox News reveals it has no idea how movies work in a very misguided tweet about Jamie Lee Curtis
                </CardTitle>,
                <CardText
                  className="articleDescription"
                  tag="p"
                >
                  On Wednesday, Fox News shared one of the dumbest takes of our time about Jamie Lee Curtis, star of the Halloween re-boot. The conservative news outlet tweeted, "Jamie Lee Curtis wields firearms in new 'Halloween' movie despite advocating for gun control." It'…
                </CardText>,
                <CardText
                  tag="p"
                >
                  <small
                    className="text-muted"
                  >
                    Mashable
                  </small>
                </CardText>,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "children": "Fox News reveals it has no idea how movies work in a very misguided tweet about Jamie Lee Curtis",
                  "tag": "h5",
                },
                "ref": null,
                "rendered": "Fox News reveals it has no idea how movies work in a very misguided tweet about Jamie Lee Curtis",
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "children": "On Wednesday, Fox News shared one of the dumbest takes of our time about Jamie Lee Curtis, star of the Halloween re-boot. The conservative news outlet tweeted, \\"Jamie Lee Curtis wields firearms in new 'Halloween' movie despite advocating for gun control.\\" It'…",
                  "className": "articleDescription",
                  "tag": "p",
                },
                "ref": null,
                "rendered": "On Wednesday, Fox News shared one of the dumbest takes of our time about Jamie Lee Curtis, star of the Halloween re-boot. The conservative news outlet tweeted, \\"Jamie Lee Curtis wields firearms in new 'Halloween' movie despite advocating for gun control.\\" It'…",
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "children": <small
                    className="text-muted"
                  >
                    Mashable
                  </small>,
                  "tag": "p",
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Mashable",
                    "className": "text-muted",
                  },
                  "ref": null,
                  "rendered": "Mashable",
                  "type": "small",
                },
                "type": [Function],
              },
            ],
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      "type": "a",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
